cylc/flow/task_pool.py

```diff
@@ -1214,70 +1216,112 @@ def spawn_on_output(self, itask, output, forced=False):
             and itask.identity not in self.expected_failed_tasks
         ):
             self.abort_task_failed = True
+
+        if not forced and output in [
+            TASK_OUTPUT_SUCCEEDED,
+            TASK_OUTPUT_EXPIRED,
+            TASK_OUTPUT_FAILED
+        ]:
+            self.remove_if_complete(itask)
+
         try:
-            children = itask.graph_children[output]
+            if forced:
+                self.tasks_to_spawn_forced[
+                    (itask, output)
```

Comment:
`(itask, output, forced)` to avoid needing two dicts?

Possibly consider using a deque or list rather than dict unless there's a need to perform `in` comparisons.